### DISCUSSION TITLE
Support for more text styles

### DISCUSSION THREAD

**Discussion by data-man** [ArthurSonzogni/FTXUI#398] [Ideas]
Thank you for the nice project!

It would be cool to have more text styles available.
E.g. this is Wezterm's styles:
https://user-images.githubusercontent.com/1754269/167286003-9916a018-8d4d-457e-9515-a31192e8c965.mp4

Many other "modern" terminal emulators also support `Overlined (SGR 53)` and various underlines.

**Comment by ArthurSonzogni** [comment]
This sounds like a good idea!
I still would like to keep only things that are reasonably supported, or the ones we can query terminal's capability and implement fallback when unsupported.
At least, I would like to ensure support for [xterm.js,](https://xtermjs.org/docs/api/vtfeatures/), kitty, gnome-terminal and Microsoft's terminal.

Curly underline looks great!

**Reply by data-man** [reply]
> 

Kitty, Alacritty, Contour, Wezterm support `4:*` underlines.
I don't know about other TE. 

According to https://github.com/microsoft/terminal/blob/main/src/renderer/atlas/AtlasEngine.h MS terminal should support dotted and double underlines:

https://github.com/microsoft/terminal/blob/dad68edc034e5e022b29881ea7e3af77b97c56a0/src/renderer/atlas/AtlasEngine.h#L395-L413

**Comment by data-man** [comment]
I hope it can be useful for your project: https://github.com/contour-terminal/endo.
https://github.com/contour-terminal/endo/blob/master/src/tui/TerminalOutput.hpp#L52-L79:
```c++
/// @brief Color representation: default, 256-color index, or true color (RGB).
using Color = std::variant<std::monostate, std::uint8_t, RgbColor>;

/// @brief Underline style for terminal output (SGR 4:n).
enum class UnderlineStyle : std::uint8_t
{
    None = 0,   ///< No underline.
    Single = 1, ///< Single underline (SGR 4:1, same as SGR 4).
    Double = 2, ///< Double underline (SGR 4:2).
    Curly = 3,  ///< Curly/wavy underline (SGR 4:3).
    Dotted = 4, ///< Dotted underline (SGR 4:4).
    Dashed = 5, ///< Dashed underline (SGR 4:5).
};

/// @brief Text styling attributes for terminal output.
struct Style
{
    Color fg;                                             ///< Foreground color.
    Color bg;                                             ///< Background color.
    bool bold = false;                                    ///< Bold text.
    bool italic = false;                                  ///< Italic text.
    bool underline = false;                               ///< Underlined text (simple SGR 4).
    bool strikethrough = false;                           ///< Strikethrough text.
    bool dim = false;                                     ///< Dim/faint text.
    bool inverse = false;                                 ///< Inverse/reverse video.
    UnderlineStyle underlineStyle = UnderlineStyle::None; ///< Extended underline style (SGR 4:n).
    Color underlineColor;                                 ///< Underline color (SGR 58;2;r;g;b or 58;5;idx).
};
```
